File: src/main/java/io/github/tanguygab/playerlistexpansion/PlayerListExpansion.java
package io.github.tanguygab.playerlistexpansion;

import java.util.*;
import java.util.logging.Level;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import io.github.tanguygab.playerlistexpansion.filters.Filter;
import io.github.tanguygab.playerlistexpansion.groups.GroupedList;
import io.github.tanguygab.playerlistexpansion.groups.ListGroup;
import io.github.tanguygab.playerlistexpansion.sorting.SortingType;
import me.clip.placeholderapi.expansion.Configurable;
import me.clip.placeholderapi.expansion.Taskable;
import org.bukkit.OfflinePlayer;

import me.clip.placeholderapi.expansion.PlaceholderExpansion;
import org.bukkit.configuration.ConfigurationSection;
import org.jetbrains.annotations.NotNull;

public class PlayerListExpansion extends PlaceholderExpansion implements Taskable, Configurable {

	private static final Pattern FORMAT_PATTERN = Pattern.compile("_format:");

	private static PlayerListExpansion instance;
	private final Map<String, PlayerList> lists = new HashMap<>();
	private final Map<String, ListGroup> listGroups = new HashMap<>();
	private final List<String> placeholders = new ArrayList<>();

	public int updateCooldown;
	public String offlineText;
	public String argumentSeparator;
	private final static Map<String, Object> defaults = new LinkedHashMap<String,Object>() {{
		put("update-cooldown", 1000);
		put("offline-text", "Offline");
		put("argument-separator", ",");
		put("lists",new HashMap<String, Map<String,Object>>() {{
			Map<String,Object> staffList = new LinkedHashMap<String, Object>() {{
				put("type","ONLINE");
				put("included",true);
				put("filters",Arrays.asList("PERMISSION:group.staff","CANSEE"));
			}};
			put("staff",staffList);
			Map<String,Object> playersList = new LinkedHashMap<String, Object>() {{
				put("type","ONLINE");
				put("included",true);
				put("filters",Arrays.asList("!PERMISSION:group.staff","CANSEE"));
			}};
			put("players",playersList);
		}});
		put("groups", new HashMap<String, Map<String,Object>>() {{
			Map<String,Object> playerlistGroup = new LinkedHashMap<String, Object>() {{
				put("gap",1);
				Map<String,Object> lists = new LinkedHashMap<String, Object>() {{
					Map<String,Object> staffList = new LinkedHashMap<String, Object>() {{
						put("title","Staff Members (%amount%)");
						put("max",10);
						put("remaining","... and %remaining% more!");
					}};
					put("staff",staffList);
					Map<String,Object> playersList = new LinkedHashMap<String, Object>() {{
						put("title","Players (%amount%)");
					}};
					put("players",playersList);
				}};
				put("lists", lists);
			}};
			put("staffAndPlayers",playerlistGroup);
		}});
	}};

	public PlayerListExpansion() {
		instance = this;
	}
	public static PlayerListExpansion get() {
		return instance;
	}

	@Override
	public @NotNull String getAuthor() {
		return "Tanguygab";
	}

	@Override
	public @NotNull String getIdentifier() {
		return "playerlist";
	}

	@Override
	public @NotNull String getVersion() {
		return "3.0.9";
	}

	@Override
	public @NotNull List<String> getPlaceholders() {
		return placeholders;
	}

	@Override
	public void start() {
		updateCooldown = getInt("update-cooldown",1000);
		offlineText = getString("offline-text","Offline");
		argumentSeparator = Pattern.quote(getString("argument-separator", ","));

		ConfigurationSection config = getConfigSection("lists");
		if (config == null) return;
		config.getValues(false).forEach((list, obj) -> {
			ConfigurationSection cfg = getConfigSection("lists." + list);
			if (cfg == null) return;

			String type0 = cfg.getString("type", "ONLINE");
			ListType type = ListType.find(type0);
			if (type == null) {
				log(Level.CONFIG, "List \""+list+"\" has an invalid type \"" + type0 + "\"! Skipping...");
				return;
			}

			boolean included = cfg.getBoolean("included", true);
			boolean duplicates = cfg.getBoolean("duplicates", true);

			List<Filter> filters = new ArrayList<>();
			cfg.getStringList("filters").forEach(filter -> {
				Filter f = Filter.find(filter);
				if (f == null) {
					log(Level.CONFIG, "List \""+list+"\" has an invalid filter type \"" + filter + "\"! Skipping...");
					return;
				}
				filters.add(f);
			});

			List<SortingType> sortingTypes = new ArrayList<>();
			cfg.getStringList("sorting-types").forEach(sortingType -> {
				SortingType s = SortingType.find(sortingType);
				if (s == null) {
					log(Level.CONFIG, "List \""+list+"\" has an invalid sorting type \"" + sortingType + "\"! Skipping...");
					return;
				}
				sortingTypes.add(s);
			});

			lists.put(list, new PlayerList(list,type,filters,sortingTypes,included,duplicates));
		});
		placeholders.addAll(lists.keySet().stream().map(listName->"%playerlist_"+listName+"_<list|amount|#>%").collect(Collectors.toList()));


		ConfigurationSection groups = getConfigSection("groups");
		if (groups == null) return;
		groups.getValues(false).forEach((group, obj) -> {
			ConfigurationSection cfg = getConfigSection("groups." + group);
			if (cfg == null) return;

			int gap = Math.max(0, cfg.getInt("gap", 1));

			ConfigurationSection listsCfg = cfg.getConfigurationSection("lists");
			if (listsCfg == null || listsCfg.getValues(false).isEmpty()) {
				log(Level.CONFIG, "Group \""+group+"\" is empty! Skipping...");
				return;
			}
			List<GroupedList> lists = new ArrayList<>();
			listsCfg.getValues(false).forEach((list,settings0) -> {
				if (!this.lists.containsKey(list)) {
					log(Level.CONFIG, "List \""+list+"\" in group \""+group+"\" doesn't exist! Skipping...");
					return;
				}

				ConfigurationSection settings = (ConfigurationSection) settings0;

				String title = settings.getString("title", list+" (%amount%)");
				int max = Math.max(0, settings.getInt("max", 0));
				String remaining = settings.getString("remaining");

				lists.add(new GroupedList(this.lists.get(list), title, max, remaining));
			});

			if (lists.isEmpty()) {
				log(Level.CONFIG, "Group \""+group+"\" is empty! Skipping...");
				return;
			}
			listGroups.put(group, new ListGroup(lists, gap));
		});
		placeholders.addAll(listGroups.keySet().stream().map(groupName->"%playerlist_group_"+groupName+"_<amount|#>%").collect(Collectors.toList()));

	}

	@Override
	public void stop() {}

	@Override
	public String onRequest(OfflinePlayer player, @NotNull String params) {
		Matcher matcher = FORMAT_PATTERN.matcher(params);
		String _params = matcher.find() ? params.substring(0, matcher.start()) : params;

		int _index = _params.lastIndexOf('_');
		if (_index == -1 || _index == params.length()-1) return null;

		String list = params.substring(0,_index);
		String output = params.substring(_index+1);

		if (list.startsWith("group_")) {
			String group = list.substring(6);
			ListGroup groupedList = listGroups.get(group);
			return groupedList != null ? groupedList.getText(player,output) : null;
		}

		PlayerList playerList = lists.get(list);
		return playerList != null ? playerList.getText(player,output) : null;
	}

	@Override
	public Map<String, Object> getDefaults() {
		return defaults;
	}

	public String getFormat(String arg, String[] args) {
		if (args.length < 2) return "{player_name}";
		String argument = arg.substring(args[0].length()+1);
		return argument.startsWith("format:") ? argument.substring(7) : "{player_name}";
	}

}

File: src/main/java/io/github/tanguygab/playerlistexpansion/PlayerListExpansion.kt
package io.github.tanguygab.playerlistexpansion

import io.github.tanguygab.playerlistexpansion.filters.Filter
import io.github.tanguygab.playerlistexpansion.groups.GroupedList
import io.github.tanguygab.playerlistexpansion.groups.ListGroup
import io.github.tanguygab.playerlistexpansion.sorting.SortingType
import me.clip.placeholderapi.expansion.Configurable
import me.clip.placeholderapi.expansion.PlaceholderExpansion
import me.clip.placeholderapi.expansion.Taskable
import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import java.util.logging.Level
import java.util.regex.Matcher
import java.util.regex.Pattern
import kotlin.math.max

class PlayerListExpansion : PlaceholderExpansion(), Taskable, Configurable {
    private val lists = HashMap<String, PlayerList>()
    private val listGroups = HashMap<String, ListGroup>()
    private val placeholders = ArrayList<String>()

    var updateCooldown: Int = 0
    lateinit var offlineText: String
    lateinit var argumentSeparator: String

    companion object {
        private val FORMAT_PATTERN: Pattern = Pattern.compile("_format:")

        private lateinit var instance: PlayerListExpansion

        private val defaults = LinkedHashMap<String, Any?>().apply {
            put("update-cooldown", 1000)
            put("offline-text", "Offline")
            put("argument-separator", ",")
            put("lists", HashMap<String, Any>().apply {
                put("staff", LinkedHashMap<String, Any>().apply {
                    put("type", "ONLINE")
                    put("included", true)
                    put("filters", listOf("PERMISSION:group.staff", "CANSEE"))
                })
                put("players", LinkedHashMap<String, Any>().apply {
                    put("type", "ONLINE")
                    put("included", true)
                    put("filters", listOf("!PERMISSION:group.staff", "CANSEE"))
                })
            })
            put("groups", HashMap<String, Any>().apply {
                put("staffAndPlayers", LinkedHashMap<String, Any>().apply {
                    put("gap", 1)
                    put("lists", LinkedHashMap<String, Any>().apply {
                        put("staff", LinkedHashMap<String, String>().apply {
                            put("title", "Staff Members (%amount%)")
                            put("max", 10)
                            put("remaining", "... and %remaining% more!")
                        })
                        put("players", LinkedHashMap<String, String>().apply {
                            put("title", "Players (%amount%)")
                        })
                    })
                })
            })
        }

        fun get(): PlayerListExpansion {
            return instance
        }
    }

    init {
        instance = this
    }

    override fun getAuthor() = "Tanguygab"
    override fun getIdentifier() = "playerlist"
    override fun getVersion() = "3.1.0"
    override fun getPlaceholders() = placeholders
    override fun getDefaults() = Companion.defaults

    override fun start() {
        updateCooldown = getInt("update-cooldown", 1000)
        offlineText = getString("offline-text", "Offline")!!
        argumentSeparator = Pattern.quote(getString("argument-separator", ","))

        val config = getConfigSection("lists")
        if (config == null) return
        config.getValues(false).forEach { (list: String, _) ->
            val cfg = getConfigSection("lists.$list")
            if (cfg == null) return@forEach

            val type0: String = cfg.getString("type", "ONLINE")!!
            val type: ListType? = ListType.find(type0)
            if (type == null) {
                log(Level.CONFIG, "List \"$list\" has an invalid type \"$type0\"! Skipping...")
                return@forEach
            }

            val included = cfg.getBoolean("included", true)
            val duplicates = cfg.getBoolean("duplicates", true)

            val filters = ArrayList<Filter>()
            cfg.getStringList("filters").forEach {
                val filter: Filter? = Filter.find(it)
                if (filter == null) {
                    log(
                        Level.CONFIG,
                        "List \"$list\" has an invalid filter type \"$it\"! Skipping..."
                    )
                    return@forEach
                }
                filters.add(filter)
            }

            val sortingTypes = ArrayList<SortingType>()
            cfg.getStringList("sorting-types").forEach { it ->
                val sortingType: SortingType? = SortingType.find(it)
                if (sortingType == null) {
                    log(
                        Level.CONFIG,
                        "List \"$list\" has an invalid sorting type \"$it\"! Skipping..."
                    )
                    return@forEach
                }
                sortingTypes.add(sortingType)
            }
            lists.put(list, PlayerList(list, type, filters, sortingTypes, included, duplicates))
        }
        placeholders.addAll(lists.keys.map { listName: String? -> "%playerlist_" + listName + "_<list|amount|#>%" })


        val groups = getConfigSection("groups")
        if (groups == null) return
        groups.getValues(false).forEach { (group: String, _) ->
            val cfg = getConfigSection("groups.$group")
            if (cfg == null) return@forEach

            val gap = max(0, cfg.getInt("gap", 1))

            val listsCfg = cfg.getConfigurationSection("lists")
            if (listsCfg == null || listsCfg.getValues(false).isEmpty()) {
                log(Level.CONFIG, "Group \"$group\" is empty! Skipping...")
                return@forEach
            }
            val lists = ArrayList<GroupedList>()
            listsCfg.getValues(false).forEach { (list: String, settings0: Any) ->
                if (!this.lists.containsKey(list)) {
                    log(Level.CONFIG, "List \"$list\" in group \"$group\" doesn't exist! Skipping...")
                    return@forEach
                }
                val settings = settings0 as ConfigurationSection

                val title: String = settings.getString("title", "$list (%amount%)")!!
                val max = max(0, settings.getInt("max", 0))
                val remaining = settings.getString("remaining")
                lists.add(GroupedList(this.lists[list]!!, title, max, remaining))
            }

            if (lists.isEmpty()) {
                log(Level.CONFIG, "Group \"$group\" is empty! Skipping...")
                return@forEach
            }
            listGroups.put(group, ListGroup(lists, gap))
        }
        placeholders.addAll(listGroups.keys.map { "%playerlist_group_${it}_<amount|#>%" })
    }

    override fun stop() {}

    override fun onRequest(player: OfflinePlayer?, params: String): String? {
        val matcher: Matcher = FORMAT_PATTERN.matcher(params)
        val uParams = if (matcher.find()) params.substring(0, matcher.start()) else params

        val uIndex = uParams.lastIndexOf('_')
        if (uIndex == -1 || uIndex == params.length - 1) return null

        val list = params.substring(0, uIndex)
        val output = params.substring(uIndex + 1)

        if (list.startsWith("group_")) {
            val group = list.substring(6)
            val groupedList = listGroups[group]
            return groupedList?.getText(player, output)
        }

        val playerList = lists[list]
        return playerList?.getText(player, output)
    }

    fun getFormat(arg: String, args: List<String>): String {
        if (args.size < 2) return "{player_name}"
        val argument = arg.substring(args[0].length + 1)
        return if (argument.startsWith("format:")) argument.substring(7) else "{player_name}"
    }
}
